// e621-api/build/subComponents/Sets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var xml2json_1 = require("../lib/xml2json");
var xmldom_1 = require("xmldom");
var parser = new xmldom_1.DOMParser();
var Sets = /** @class */ (function () {
    function Sets(pageLimit, requestServices) {
        this.pageLimit = pageLimit;
        this.requestServices = requestServices;
    }
    /** List most recent sets, limit of 50 per page
     *
     * **PLEASE NOTE**: This is being converted from XML as the JSON endpoint is 30x slower than the
     * XML endpoint, so it's faster to convert them. There may also be minor bugs even when custom converting
     * @param {number} [page] Page number to return
     * @returns An array of set data (without post info, use `showSet()` method to retrive the post info)
     * @memberof Sets
     */
    Sets.prototype.getAllSets = function (page) {
        var _this = this;
        var url = "https://e621.net/set/index.xml?limit=50";
        if (page)
            url += "&page=" + page;
        return this.requestServices.get(url)
            .then(function (response) {
            // format the XML string to JSON
            var document = parser.parseFromString(response, 'text/xml');
            var json = xml2json_1.xmlToJson(document);
            // clean the conversion artifacts
            var cleanedResponse = _this.beautifySetJSONArray(json.sets.set);
            return cleanedResponse;
        })
            .catch(function (err) {
            throw Error(err);
        });
    };
    /** Get sets that include the given post's ID
     *
     * **PLEASE NOTE**: This is being converted from XML as the JSON endpoint is 30x slower than the
     * XML endpoint, so it's faster to convert them. There may also be minor bugs even when custom converting
     * @param {number} postID Post to filter sets by
     * @param {number} [page] The page number to return, if there is only one page this is ignored
     * @returns An array of set data (without post info, use `showSet()` method to retrive the post info)
     * @memberof Sets
     */
    Sets.prototype.getSetsByPostID = function (postID, page) {
        var _this = this;
        var url = "https://e621.net/set/index.xml?limit=50&post_id=" + postID;
        if (page)
            url += "&page=" + page;
        return this.requestServices.get(url)
            .then(function (response) {
            // format the XML string to JSON
            var document = parser.parseFromString(response, 'text/xml');
            var json = xml2json_1.xmlToJson(document);
            // clean the conversion artifacts
            var cleanedResponse = _this.beautifySetJSONArray(json.sets.set);
            return cleanedResponse;
        })
            .catch(function (err) {
            throw Error(err);
        });
    };
    /** Get sets created by the given user's ID
     *
     * **PLEASE NOTE**: This is being converted from XML as the JSON endpoint is 30x slower than the
     * XML endpoint, so it's faster to convert them. There may also be minor bugs even when custom converting
     * @param {number} userID User ID to filter sets by
     * @param {number} [page] The page number to return, if there is only one page this is ignored
     * @returns An array of set data (without post info, use `showSet()` method to retrive the post info)
     * @memberof Sets
     */
    Sets.prototype.getSetsByUserID = function (userID, page) {
        var _this = this;
        var url = "https://e621.net/set/index.xml?user_id=" + userID;
        if (page)
            url += "&page=" + page;
        return this.requestServices.get(url)
            .then(function (response) {
            // format the XML string to JSON
            var document = parser.parseFromString(response, 'text/xml');
            var json = xml2json_1.xmlToJson(document);
            // clean the conversion artifacts
            var cleanedResponse = _this.beautifySetJSONArray(json.sets.set);
            return cleanedResponse;
        })
            .catch(function (err) {
            throw Error(err);
        });
    };
    /** Get a set by its `setID`
     *
     * **PLEASE NOTE**: This is being converted from XML as the JSON endpoint is 30x slower than the
     * XML endpoint, so it's faster to convert them. There may also be minor bugs even when custom converting
     * @param {(number | string)} setID ID of the set to convert and retrieve
     * @returns Promise<e621SetJSONConvertedWithPosts> - a single converted XML set
     * @memberof Sets
     */
    Sets.prototype.showSet = function (setID) {
        var _this = this;
        var url = "https://e621.net/set/show.xml?id=" + setID;
        return this.requestServices.get(url)
            .then(function (response) {
            // format the XML string to JSON
            var document = parser.parseFromString(response, 'text/xml');
            var json = xml2json_1.xmlToJson(document);
            // clean the conversion artifacts
            // this is going to take a massive amount of work
            var cleanedSet = _this.beautifySetJSONSingle(json["post-set"], json["post-set"].posts.post);
            return cleanedSet;
        })
            .catch(function (err) {
            throw Error(err);
        });
    };
    /** Cretae a new empty set
     * @param {string} setName Name of the set to create
     * @param {string} shortName SHort named of the set
     * @param {string} setDescription Description for the set
     * @param {boolean} isPublic If the set will be public
     * @param {boolean} transferOnDelete Whether to replace deleted posts with their parents
     * @returns Promise<e621POSTResponse>
     * @memberof Sets
     */
    Sets.prototype.create = function (setName, shortName, setDescription, isPublic, transferOnDelete) {
        var url = "https://e621.net/set/create.json";
        var postObj = {
            "set[name]": setName,
            "set[shortname]": shortName,
            "set[description]": setDescription,
            "set[public]": isPublic,
            "set[transfer_to_parent_on_delete]": transferOnDelete
        };
        return this.requestServices.post(url, postObj)
            .then(function (response) {
            return response;
        })
            .catch(function (err) {
            throw Error(err);
        });
    };
    /** Update a set's name by ID
     * @param {(number | string)} setID ID of the set to update
     * @param {string} newName New name of the set
     * @returns Promise<e621POSTResponse>
     * @memberof Sets
     */
    Sets.prototype.updateName = function (setID, newName) {
        var url = "https://e621.net/set/update.json";
        var postObj = {
            "set[id]": setID,
            "set[name]": newName,
        };
        return this.requestServices.post(url, postObj)
            .then(function (response) {
            return response;
        })
            .catch(function (err) {
            throw Error(err);
        });
    };
    /** Update a set's shortname by ID
     * @param {(number | string)} setID ID of the set to update
     * @param {string} newShortName New shortname of the set
     * @returns Promise<e621POSTResponse>
     * @memberof Sets
     */
    Sets.prototype.updateShortName = function (setID, newShortName) {
        var url = "https://e621.net/set/update.json";
        var postObj = {
            "set[id]": setID,
            "set[shortname]": newShortName,
        };
        return this.requestServices.post(url, postObj)
            .then(function (response) {
            return response;
        })
            .catch(function (err) {
            throw Error(err);
        });
    };
    /** Update a set's description by its ID
     * @param {(number | string)} setID ID of the set to update
     * @param {string} newDescription New description of the set
     * @returns Promise<e621POSTResponse>
     * @memberof Sets
     */
    Sets.prototype.updateDescription = function (setID, newDescription) {
        var url = "https://e621.net/set/update.json";
        var postObj = {
            "set[id]": setID,
            "set[description]": newDescription,
        };
        return this.requestServices.post(url, postObj)
            .then(function (response) {
            return response;
        })
            .catch(function (err) {
            throw Error(err);
        });
    };
    /** Update a set's public status by its `setID`
     * @param {(number | string)} setID ID of the set to update
     * @param {string} isPublic Public status of the set
     * @returns Promise<e621POSTResponse>
     * @memberof Sets
     */
    Sets.prototype.updatePublicStatus = function (setID, isPublic) {
        var url = "https://e621.net/set/update.json";
        var postObj = {
            "set[id]": setID,
            "set[pulibc]": isPublic,
        };
        return this.requestServices.post(url, postObj)
            .then(function (response) {
            return response;
        })
            .catch(function (err) {
            throw Error(err);
        });
    };
    /** Update a set's transferOnDelete status by its `setID`
     * @param {(number | string)} setID ID of the set to update
     * @param {string} transferOnDelete Public status of the set
     * @returns Promise<e621POSTResponse>
     * @memberof Sets
     */
    Sets.prototype.updateTransferOnDeleteStatus = function (setID, transferOnDelete) {
        var url = "https://e621.net/set/update.json";
        var postObj = {
            "set[id]": setID,
            "set[transfer_to_parent_on_delete]": transferOnDelete,
        };
        return this.requestServices.post(url, postObj)
            .then(function (response) {
            return response;
        })
            .catch(function (err) {
            throw Error(err);
        });
    };
    /** Add a post to a set
     * @param {(number | string)} setID ID of the set to add the post to
     * @param {number} postID ID of the post to add to the set
     * @returns Promise<e621POSTResponse>
     * @memberof Sets
     */
    Sets.prototype.addPost = function (setID, postID) {
        var url = "https://e621.net/set/add_post.json";
        var postObj = {
            "set_id": setID,
            "post_id": postID
        };
        return this.requestServices.post(url, postObj)
            .then(function (response) {
            return response;
        })
            .catch(function (err) {
            throw Error(err);
        });
    };
    /** Remove a post from a set
     * @param {(number | string)} setID ID of the set to remove the post from
     * @param {number} postID ID of the post to remove from the set
     * @returns Promise<e621POSTResponse>
     * @memberof Sets
     */
    Sets.prototype.removePost = function (setID, postID) {
        var url = "https://e621.net/set/remove_post.json";
        var postObj = {
            "set_id": setID,
            "post_id": postID
        };
        return this.requestServices.post(url, postObj)
            .then(function (response) {
            return response;
        })
            .catch(function (err) {
            throw Error(err);
        });
    };
    /** Delete a set (If you have proper permissions)
     * @param {(number | string)} setID ID of the set to deleteF
     * @returns Promise<e621POSTResponse>
     * @memberof Sets
     */
    Sets.prototype.destroy = function (setID) {
        var url = "https://e621.net/set/destroy.json";
        var postObj = {
            "id": setID,
        };
        return this.requestServices.post(url, postObj)
            .then(function (response) {
            return response;
        })
            .catch(function (err) {
            throw Error(err);
        });
    };
    Sets.prototype.beautifySetJSONArray = function (convertedSetArray) {
        var arrayToReturn = [];
        convertedSetArray.forEach(function (setData) {
            // coerce the JSON into a custom object
            var cleanedObject = {};
            cleanedObject.id = parseInt(setData.id);
            cleanedObject.name = setData.name;
            cleanedObject.created_at = setData["created-at"];
            cleanedObject.updated_at = setData["updated-at"];
            cleanedObject.user_id = parseInt(setData["user-id"]);
            cleanedObject.description = setData.description;
            cleanedObject.shortname = setData.shortname;
            cleanedObject.post_count = parseInt(setData["post-count"]);
            cleanedObject.public = JSON.parse(setData.public);
            cleanedObject.transfer_to_parent_on_delete = JSON.parse(setData["transfer-to-parent-on-delete"]);
            arrayToReturn.push(cleanedObject);
        });
        return arrayToReturn;
    };
    Sets.prototype.beautifySetJSONSingle = function (convertedSetJSON, convertedPostJSON) {
        // coerce the JSON into a custom object
        var cleanedObject = {};
        // array to hold all of the posts we've cleaned to have their correct typings
        var cleanedPosts = [];
        cleanedObject.id = parseInt(convertedSetJSON.id);
        cleanedObject.name = convertedSetJSON.name;
        cleanedObject.created_at = convertedSetJSON["created-at"];
        cleanedObject.updated_at = convertedSetJSON["updated-at"];
        cleanedObject.user_id = parseInt(convertedSetJSON["user-id"]);
        cleanedObject.description = convertedSetJSON.description;
        cleanedObject.shortname = convertedSetJSON.shortname;
        cleanedObject.post_count = parseInt(convertedSetJSON["post-count"]);
        cleanedObject.public = JSON.parse(convertedSetJSON.public);
        cleanedObject.transfer_to_parent_on_delete = JSON.parse(convertedSetJSON["transfer-to-parent-on-delete"]);
        convertedPostJSON.forEach(function (post, index) {
            var cleanedPost = {
                id: parseInt(post.id),
                tags: post.tags,
                locked_tags: post.locked_tags,
                description: post.description,
                created_at: post.created_at,
                creator_id: parseInt(post.creator_id),
                author: post.author,
                change: parseInt(post.change),
                source: post.source,
                score: parseInt(post.score),
                fav_count: parseInt(post.fav_count),
                md5: post.md5,
                file_size: parseInt(post.file_size),
                file_url: post.file_url,
                file_ext: post.file_ext,
                preview_url: post.preview_url,
                preview_width: parseInt(post.preview_width),
                preview_height: parseInt(post.preview_height),
                sample_url: post.sample_url,
                sample_width: parseInt(post.sample_width),
                sample_height: parseInt(post.sample_height),
                rating: post.rating,
                status: post.status,
                width: parseInt(post.width),
                height: parseInt(post.height),
                has_comments: JSON.parse(post.has_comments),
                has_notes: JSON.parse(post.has_notes),
                has_children: JSON.parse(post.has_children),
                children: post.children,
                parent_id: null,
            };
            // check if a parent ID isn't some garbage XML that means nothing
            if (isNaN(parseInt(post.parent_id)) == false) {
                cleanedPost.parent_id = post.parent_id;
            }
            // make sure the artist prop deson't return XML garbage
            if (post.artist.hasOwnProperty('artist')) {
                cleanedPost.artist = post.artist.artist;
            }
            else {
                cleanedPost.artist = post.artist;
            }
            // if > 1 source, handle the XML -> JSON array conversion
            if (post.sources) {
                if (post.sources.hasOwnProperty('source')) {
                    cleanedPost.sources = post.sources.source;
                }
            }
            else {
                cleanedPost.sources = post.sources;
            }
            cleanedPosts.push(cleanedPost);
        });
        cleanedObject.posts = cleanedPosts;
        return cleanedObject;
    };
    return Sets;
}());
exports.default = Sets;
