// e621-api/build/e621-api.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// Imports and dependencies
var Tags_1 = __importDefault(require("./subComponents/Tags"));
var Posts_1 = __importDefault(require("./subComponents/Posts"));
var Artists_1 = __importDefault(require("./subComponents/Artists"));
var Comments_1 = __importDefault(require("./subComponents/Comments"));
var Pools_1 = __importDefault(require("./subComponents/Pools"));
var Users_1 = __importDefault(require("./subComponents/Users"));
var Blips_1 = __importDefault(require("./subComponents/Blips"));
var Sets_1 = __importDefault(require("./subComponents/Sets"));
var Dmail_1 = __importDefault(require("./subComponents/Dmail"));
var Notes_1 = __importDefault(require("./subComponents/Notes"));
var Wiki_1 = __importDefault(require("./subComponents/Wiki"));
var Forums_1 = __importDefault(require("./subComponents/Forums"));
var RequestService_1 = require("./RequestService");
var e621 = /** @class */ (function () {
    /**
     * Creates an instance of e621.
     * @param {string} userAgent This identifies you to the API
     * @param {string} [userName] Your user name (optional, if making changes to data)
     * @param {string} [apiKey] Your API key to use (optional, if making changes to data)
     * @param {number} [pageLimit] Page limit for paginated calls from e621
     * @memberof e621
     */
    function e621(userAgent, userName, apiKey, pageLimit) {
        this.userAgent = userAgent;
        if (pageLimit)
            this.pageLimit = pageLimit;
        else
            this.pageLimit = 3;
        if (userName && apiKey) {
            this.requestservices = new RequestService_1.RequestServices(userAgent, userName, apiKey);
        }
        else {
            // we won't send the RequestServices a username + hash if they are not given
            this.requestservices = new RequestService_1.RequestServices(userAgent);
        }
        // assign the properties as new classes with the page limit and the methods for get/post 
        // operations as arguments
        this.tags = new Tags_1.default(this.pageLimit, this.requestservices);
        this.posts = new Posts_1.default(this.pageLimit, this.requestservices);
        this.artists = new Artists_1.default(this.pageLimit, this.requestservices);
        this.comments = new Comments_1.default(this.pageLimit, this.requestservices);
        this.pools = new Pools_1.default(this.pageLimit, this.requestservices);
        this.users = new Users_1.default(this.pageLimit, this.requestservices);
        this.blips = new Blips_1.default(this.pageLimit, this.requestservices);
        this.sets = new Sets_1.default(this.pageLimit, this.requestservices);
        this.dmail = new Dmail_1.default(this.pageLimit, this.requestservices);
        this.notes = new Notes_1.default(this.pageLimit, this.requestservices);
        this.wiki = new Wiki_1.default(this.pageLimit, this.requestservices);
        this.forums = new Forums_1.default(this.pageLimit, this.requestservices);
    }
    Object.defineProperty(e621.prototype, "agent", {
        get: function () { return this.userAgent; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(e621.prototype, "limit", {
        get: function () { return this.pageLimit; },
        enumerable: true,
        configurable: true
    });
    return e621;
}());
exports.default = e621;
